### PR TITLE
Add GitHub Actions CI

### PR DESCRIPTION
**Pull request by slabasan**
Add CI workflow using GitHub Actions
- coverage across Ubuntu 18, python >= 3.6
- validate both C and Python binding tests

Additional changes:
- remove py extension from C smoke test
- add build status badge to README
- add softlink to README for setup.py

Updating PR title, but keeping original branch name so we do not have to start a new PR (originally this PR was targeting Azure CI).

### CODE REVIEW

**Comment by dongahn** [comment]
@slabasan: Oh ... so fast! I will add @tpatki as a reviewer and also tag @zekemorton, in case he's interest. 

Also please add some description into the PR. You could also add a bit more details on the commit -- what's the coverage that this commit includes etc. 

Thanks!

**Comment by dongahn** [comment]
@zekemorton: are you well verse with GitHub Action to give a quick review for this PR when @slabasan is ready?

**Comment by slabasan** [comment]
Got it, will do! I was hoping this repo might launch this pipeline on azure, so I could see if my paths are set correctly, I have the coverage I want, etc. But I'm not sure if something in the settings needs to be modified to allow Azure to have access. I might add a github actions file to this PR and see if workflow runs.

**Comment by dongahn** [comment]
I assume you only include cbinding test just to remain simple. And you will add python binding later, correct? https://github.com/flux-framework/PerfFlowAspect/blob/main/src/python/test/t0001-pybinding-basic.t

**Comment by tpatki** [comment]
@slabasan I don't see the python tests in here, are you adding those later?

**Comment by slabasan** [comment]
Yes that's correct, this PR is not done yet. Wanted to start with a small workflow to get the paths right, and then I'll expand to the other tests.

**Comment by slabasan** [comment]
@tpatki or @dongahn Can you grant access in this repo for Azure and/or Github Actions using Settings? This should enable the workflows to run.

**Comment by tpatki** [comment]
@slabasan I couldn't find an option specific to Azure or Github Actions under the Actions tab. There are some options for outside collaborators, which may impact this setting. I added you as a collaborator with admin rights -- not sure if that helps you change the relevant settings.

**Comment by dongahn** [comment]
@grondo: a quick question. What setting should be set to allow @slabasan to run her CI workflow through GitHub Action?

**Comment by slabasan** [comment]
I am thinking that the CI yaml file needs to be in main first before it will run..not ideal for testing the workflow before merging.

**Comment by grondo** [comment]
@slabasan - you should be able to enable workflows on your personal fork of this repo and test the CI there (to check results look under your fork and check the Actions tab).

I don't remember if actions run by default when they are _added_ in a PR. It seems like they should, o/w you risk merging a broken workflow. I'll do some research.

**Comment by garlick** [comment]
I ran into this recently over in https://github.com/chaos/diod and I concluded that you needed to merge a "dummy" workflow to the main branch before you could put up the real one as a PR and find out if it works, e.g. something like this:
```
name: ci

on: [ workflow_dispatch ]

jobs:
  dummyjob:
    steps:
    - name: dummystep
```

**Comment by grondo** [comment]
> you should be able to enable workflows on your personal fork of this repo and test the CI there

I just pushed a copy of your branch (with the update to run workflow on any `push`) to my fork and the workflow ran (no configuration needed):

https://github.com/grondo/PerfFlowAspect/actions/runs/1440660178

**Comment by slabasan** [comment]
Thanks @grondo and @garlick for the tips!

**Comment by dongahn** [comment]
Yeah! Thanks everyone!

**Comment by dongahn** [comment]
@slabasan: I see you started to add different LLVM versions. Currently, the C++ annotation support depends on older versions of LLVM/Clang and porting it to different versions is future item. Probably shouldn't be a part of this PR.

**Comment by dongahn** [comment]
https://github.com/flux-framework/PerfFlowAspect/blob/main/src/c/weaver/weave/perfflow_weave.cpp is the main file that use LLVM APIs. From Laguna, porting this to more up to date LLVM versions should be relatively straightforward.

**Comment by slabasan** [comment]
@dongahn Thanks! I am not intending for the CI to cover different LLVM versions, just trying to find an LLVM version that will build everything successfully for Ubuntu 20. So far, none of the packages I have tried results in a successful build of the test directory, and the error messages have not been useful (just giving me a segmentation fault, can't run this executable).

**Comment by dongahn** [comment]
With the older ubuntu, `apt install clang llvm-dev` works? Perhaps we just have coverage for the older ubuntu. And then when we upgrade LLVM API, we add newer versions of ubuntu?

**Comment by slabasan** [comment]
Oh good point, I was using `ubuntu-latest` instead of `ubuntu-18.04`, let me switch over to the older version.

File: .github/workflows/github-actions.yml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+      - name: Build Weaver
+        run: |
+          cd src/c/weaver/build
+          make VERBOSE=1
+
+      - name: Run C Smoke Tests
+        run: |
+          cd src/c/test
+          make
+          sh t0001-cbinding-basic.t
```
Comment (dongahn) [inline .github/workflows/github-actions.yml:46]: `t0001-cbinding-basic.t` already include `#! /bin/sh`. Is there a specific reason to include `sh` additionally?

Just so that I understand, if the script execution return non zero, this step will fail in the CI?

File: .github/workflows/github-actions.yml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      - name: Install Dependencies
+        run: |
+          python -m pip install --upgrade pip pytest setuptools
+          python src/python/setup.py install
+          python -m pip list
+
+      - name: Run Python Smoke Tests
+        run: |
+          cd src/python/test
+          sh t0001-pybinding-basic.t
```
Comment (dongahn) [inline .github/workflows/github-actions.yml:73]: Same comment as above.

File: README.md [hunk 3]
```diff
@@ -2,6 +2,8 @@
 are not yet stable.* The github issue tracker is the primary way to
 communicate with the developers.
 
+[![Build Status](https://github.com/flux-framework/PerfFlowAspect/actions/workflows/github-actions.yaml/badge.svg)](https://github.com/flux-framework/PerfFlowAspect/actions)
```
Comment (dongahn) [inline README.md:5]: I think this can be its own commit. (following flux code-development convention -- having each topical transaction as its own commit).

File: src/c/test/t0001-cbinding-basic.t [hunk 4]
```diff
@@ -126,7 +126,7 @@ PERFFLOW_OPTIONS="log-filename-include=instance-path" ../smoketest &&
 
 test_expect_success 'PERFFLOW_OPTIONS: Flux f58-encoded jobid supported' '
     FLUX_JOB_ID=ƒeF9QZG3 \
-PERFFLOW_OPTIONS="log-filename-include=instance-path" ../smoketest.py &&
+PERFFLOW_OPTIONS="log-filename-include=instance-path" ../smoketest &&
```
Comment (dongahn) [inline src/c/test/t0001-cbinding-basic.t:129]: This should be its down commit as well.

File: src/python/setup.py [hunk 5]
```diff
@@ -35,7 +35,7 @@ def load_readme():
     package_dir = {
         '': 'src/python'
         },
-    packages=['perfflowaspect']
+    packages=['perfflowaspect'],
```
Comment (dongahn) [inline src/python/setup.py:38]: This should be its own commit.

File: src/python/README.md [hunk 6]
```diff
@@ -0,0 +1 @@
+../../README.md
```
Comment (dongahn) [inline src/python/README.md:1]: This should be its own commit.

**Review by dongahn — APPROVED** [review]
This looks great and thank you for brining up to speed so quickly.

I give my approval. However, it would be nice to address the suggestions -- in particular making each topical change as a distinct commit.

**Comment by slabasan** [comment]
Thanks @dongahn! I will get these comments addressed, today and will push them up as individual PRs.

File: .github/workflows/github-actions.yml (hunk 1 above)
Comment (slabasan) [reply .github/workflows/github-actions.yml]: Good catch, no we do not need to run this script using sh explicitly from the yaml file, we can just launch it with `./` and it behaves just the same. Updated in this PR.

File: .github/workflows/github-actions.yml (hunk 2 above)
Comment (slabasan) [reply .github/workflows/github-actions.yml]: Removed extraneous `sh` call

**Comment by slabasan** [comment]
Ok @dongahn, this is ready for 1 last check.

**Comment by dongahn** [comment]
Perfect!
